# Software License Agreement (BSD License)
# Copyright Â© 2022-2023 belongs to Shadow Robot Company Ltd.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without modification,
# are permitted provided that the following conditions are met:
#   1. Redistributions of source code must retain the above copyright notice,
#      this list of conditions and the following disclaimer.
#   2. Redistributions in binary form must reproduce the above copyright notice,
#      this list of conditions and the following disclaimer in the documentation
#      and/or other materials provided with the distribution.
#   3. Neither the name of Shadow Robot Company Ltd nor the names of its contributors
#      may be used to endorse or promote products derived from this software without
#      specific prior written permission.
#
# This software is provided by Shadow Robot Company Ltd "as is" and any express
# or implied warranties, including, but not limited to, the implied warranties of
# merchantability and fitness for a particular purpose are disclaimed. In no event
# shall the copyright holder be liable for any direct, indirect, incidental, special,
# exemplary, or consequential damages (including, but not limited to, procurement of
# substitute goods or services; loss of use, data, or profits; or business interruption)
# however caused and on any theory of liability, whether in contract, strict liability,
# or tort (including negligence or otherwise) arising in any way out of the use of this
# software, even if advised of the possibility of such damage.

# Minimum alllowed range for the joints in this particular script
min_range: {"THJ2": -40, "THJ3": -12, "THJ4": 0, "THJ5": -55,
             "FFJ1": 0, "FFJ2": 20, "FFJ3": 0, "FFJ4": -20,
             "MFJ1": 0, "MFJ2": 20, "MFJ3": 0, "MFJ4": -10,
             "RFJ1": 0, "RFJ2": 20, "RFJ3": 0, "RFJ4": -10,
             "LFJ1": 10, "LFJ2": 10, "LFJ3": 0, "LFJ4": -20, "LFJ5": 0,
             "WRJ1": -20, "WRJ2": -10}

# Maximum alllowed range for the joints in this particular script
max_range: {"THJ2": 20, "THJ3": 12, "THJ4": 70, "THJ5": 0,
             "FFJ1": 20, "FFJ2": 90, "FFJ3": 90, "FFJ4": 0,
             "MFJ1": 20, "MFJ2": 90, "MFJ3": 90, "MFJ4": 0,
             "RFJ1": 20, "RFJ2": 90, "RFJ3": 90, "RFJ4": 0,
             "LFJ1": 20, "LFJ2": 90, "LFJ3": 90, "LFJ4": 0, "LFJ5": 1,
             "WRJ1": 10, "WRJ2": 5}

####################
# POSE DEFINITIONS #
####################
count_down_3: {"FFJ4": -12,
               "RFJ4": -12,
               "LFJ1": 85, "LFJ2": 85, "LFJ3": 85,
               "THJ2": 40}

count_down_2: {"MFJ4": 12,
               "RFJ1": 85, "RFJ2": 85, "RFJ3": 85, "RFJ4": 0}

count_down_1: {"FFJ4": 0,
               "MFJ1": 85, "MFJ2": 85, "MFJ3": 85, "MFJ4": 0}

count_down_0_ff: {"FFJ1": 85, "FFJ2": 85, "FFJ3": 85}
count_down_0_th: {"THJ2": 35, "THJ4": 35}

# scissiors position
scissors: {"FFJ1": 0, "FFJ2": 0, "FFJ3": 0, "FFJ4": -12,
           "MFJ1": 0, "MFJ2": 0, "MFJ3":  0, "MFJ4": 0.12,
           "RFJ1": 85, "RFJ2": 85, "RFJ3": 85, "RFJ4": 0, 
           "LFJ1": 85, "LFJ2": 85, "LFJ3": 85, "LFJ4": 0, "LFJ5": 0, 
           "THJ1": 0, "THJ2": 40, "THJ3": 0, "THJ4": 0, "THJ5": 0,
           "WRJ1": 0, "WRJ2": 0}
rock: {"FFJ1": 85, "FFJ2": 85, "FFJ3": 85, "FFJ4": 0,
       "MFJ1": 85, "MFJ2": 85, "MFJ3": 85, "MFJ4": 0,
       "RFJ1": 85, "RFJ2": 85, "RFJ3": 85, "RFJ4": 0, 
       "LFJ1": 85, "LFJ2": 85, "LFJ3": 85, "LFJ4": 0, "LFJ5": 0, 
       "THJ1": 40, "THJ2": 40, "THJ3": 5, "THJ4": 71, "THJ5": 10,
       "WRJ1": 0, "WRJ2": 0}
paper: {"FFJ1": 0, "FFJ2": 0, "FFJ3": 0, "FFJ4": 0,
        "MFJ1": 0, "MFJ2": 0, "MFJ3":  0, "MFJ4": 0,
        "RFJ1": 0, "RFJ2": 0, "RFJ3": 0, "RFJ4": 0, 
        "LFJ1": 0, "LFJ2": 0, "LFJ3": 0, "LFJ4": 0, "LFJ5": 0, 
        "THJ1": 0, "THJ2": 40, "THJ3": 0, "THJ4": 0, "THJ5": 0,
        "WRJ1": 0, "WRJ2": 0}

# Pregrasp position for the Hand
pregrasp_pos: {"THJ2": 12, "THJ3": 15, "THJ4": 69, "THJ5": -23,
                "FFJ1": 0, "FFJ2": 40, "FFJ3": 21, "FFJ4": -15,
                "MFJ1": 0, "MFJ2": 40, "MFJ3": 21, "MFJ4": 0,
                "RFJ1": 0, "RFJ2": 40, "RFJ3": 21, "RFJ4": -7,
                "LFJ1": 0, "LFJ2": 40, "LFJ3": 21, "LFJ4": -10, "LFJ5": 0,
                "WRJ1": -30}

# Grasp position for the Hand
grasp_pos: {"FFJ1": 0, "FFJ2": 80, "FFJ3": 47, "FFJ4": -19,
            "LFJ1": 0, "LFJ2": 77, "LFJ3": 51, "LFJ4": -19, "LFJ5": 0,
            "MFJ1": 0, "MFJ2": 76, "MFJ3": 52, "MFJ4": 0,
            "RFJ1": 0, "RFJ2": 78, "RFJ3": 53., "RFJ4": -18,
            "THJ1": 24, "THJ2": 30, "THJ3": 12, "THJ4": 69, "THJ5": -3}

# Random position for the Hand (initialied at 0)
rand_pos: {"THJ2": 0, "THJ3": 0, "THJ4": 0, "THJ5": 0,
            "FFJ1": 0, "FFJ2": 0, "FFJ3": 0, "FFJ4": 0,
            "MFJ1": 0, "MFJ2": 0, "MFJ3": 0, "MFJ4": 0,
            "RFJ1": 0, "RFJ2": 0, "RFJ3": 0, "RFJ4": 0,
            "LFJ1": 0, "LFJ2": 0, "LFJ3": 0, "LFJ4": 0, "LFJ5": 0,
            "WRJ1": 0, "WRJ2": 0}
# flex first finger
flex_ff: {"FFJ1": 80, "FFJ2": 80, "FFJ3": 80, "FFJ4": 0}
# extend first finger
ext_ff: {"FFJ1": 0, "FFJ2": 0, "FFJ3": 0, "FFJ4": 0}
# flex middle finger
flex_mf: {"MFJ1": 80, "MFJ2": 80, "MFJ3": 80, "MFJ4": 0}
# extend middle finger
ext_mf: {"MFJ1": 0, "MFJ2": 0, "MFJ3": 0, "MFJ4": 0}
# flex ring finger
flex_rf: {"RFJ1": 80, "RFJ2": 80, "RFJ3": 80, "RFJ4": 0}
# extend ring finger
ext_rf: {"RFJ1": 0, "RFJ2": 0, "RFJ3": 0, "RFJ4": 0}
# flex little finger
flex_lf: {"LFJ1": 80, "LFJ2": 80, "LFJ3": 80, "LFJ4": 0}
# extend middle finger
ext_lf: {"LFJ1": 0, "LFJ2": 0, "LFJ3": 0, "LFJ4": 0}
# flex thumb step 1
flex_th_1: {"THJ1": 0, "THJ2": 0, "THJ3": 0, "THJ4": 70, "THJ5": 0}
# flex thumb step 2
flex_th_2: {"THJ1": 35, "THJ2": 38, "THJ3": 10, "THJ4": 70, "THJ5": 47}
# extend thumb step 1
ext_th_1: {"THJ1": 90, "THJ2": -40, "THJ3": -10, "THJ4": 35, "THJ5": -60}
# extend thumb step 2
ext_th_2: {"THJ1": 0, "THJ2": 0, "THJ3": 0, "THJ4": 0, "THJ5": 0}
# zero thumb
zero_th: {"THJ1": 0, "THJ2": 0, "THJ3": 0, "THJ4": 0, "THJ5": 0}
# Pre O.K. with first finger
pre_ff_ok: {"THJ4": 70}
# O.K. with first finger
ff_ok: {"THJ1": 25, "THJ2": 11, "THJ3": 0, "THJ4": 53, "THJ5": 14,
         "FFJ1": 0, "FFJ2": 75, "FFJ3": 37, "FFJ4": 6,
         "MFJ1": 0, "MFJ2": 42, "MFJ3": 33, "MFJ4": 9,
         "RFJ1": 0, "RFJ2": 50, "RFJ3": 18, "RFJ4": -7,
         "LFJ1": 0, "LFJ2": 30, "LFJ3": 0, "LFJ4": -6, "LFJ5": 7}
# O.K. transition from first finger to middle finger
ff2mf_ok: {"THJ1": 5, "THJ2": 12, "THJ3": 4, "THJ4": 60, "THJ5": 2,
            "FFJ1": 0, "FFJ2": 14, "FFJ3": 7, "FFJ4": -0.4,
            "MFJ1": 0, "MFJ2": 42, "MFJ3": 33, "MFJ4": 8,
            "RFJ1": 0, "RFJ2": 50, "RFJ3": 18, "RFJ4": -9,
            "LFJ1": 0, "LFJ2": 30, "LFJ3": 0, "LFJ4": -12, "LFJ5": 4}
# O.K. with middle finger
mf_ok: {"THJ1": 7, "THJ2": 24, "THJ3": 0, "THJ4": 62, "THJ5": 29,
         "FFJ1": 0, "FFJ2": 15, "FFJ3": 7, "FFJ4": -0.4,
         "MFJ1": 5, "MFJ2": 85, "MFJ3": 24, "MFJ4": 8,
         "RFJ1": 0, "RFJ2": 50, "RFJ3": 18, "RFJ4": -10,
         "LFJ1": 0, "LFJ2": 30, "LFJ3": 0, "LFJ4": -12, "LFJ5": 4}
# O.K. transition from middle finger to ring finger
mf2rf_ok: {"THJ1": 5, "THJ2": -5, "THJ3": 15, "THJ4": 70, "THJ5": 15,
            "FFJ1": 0, "FFJ2": 14, "FFJ3": 7, "FFJ4": -0.4,
            "MFJ1": 0, "MFJ2": 40, "MFJ3": 4, "MFJ4": -1,
            "RFJ1": 0, "RFJ2": 50, "RFJ3": 18, "RFJ4": -10,
            "LFJ1": 0, "LFJ2": 30, "LFJ3": 0, "LFJ4": -12, "LFJ5": 4}
# O.K. with ring finger
rf_ok: {"THJ1": 3, "THJ2": 19, "THJ3": 0, "THJ4": 70, "THJ5": 47,
         "FFJ1": 0, "FFJ2": 14, "FFJ3": 7, "FFJ4": -0.4,
         "MFJ1": 0, "MFJ2": 45, "MFJ3": 4, "MFJ4": -1,
         "RFJ1": 3, "RFJ2": 85, "RFJ3": 28, "RFJ4": -9,
         "LFJ1": 0, "LFJ2": 30, "LFJ3": 0, "LFJ4": -12, "LFJ5": 4}
# O.K. transition from ring finger to little finger
rf2lf_ok: {"THJ1": 5, "THJ2": 4.5, "THJ3": 8, "THJ4": 60, "THJ5": 17,
            "FFJ1": 0, "FFJ2": 14, "FFJ3": 7, "FFJ4": -0.4,
            "MFJ1": 0, "MFJ2": 45, "MFJ3": 4, "MFJ4": -1,
            "RFJ1": 0, "RFJ2": 30, "RFJ3": 6, "RFJ4": 0.5,
            "LFJ1": 0, "LFJ2": 30, "LFJ3": 0, "LFJ4": -12, "LFJ5": 4}
# O.K. with little finger
lf_ok: {"THJ1": 5, "THJ2": 14, "THJ3": 10, "THJ4": 69, "THJ5": 15,
         "FFJ1": 0, "FFJ2": 14, "FFJ3": 7, "FFJ4": -0.4,
         "MFJ1": 0, "MFJ2": 15, "MFJ3": 4, "MFJ4": -1,
         "RFJ1": 0, "RFJ2": 15, "RFJ3": 6, "RFJ4": 0.5,
         "LFJ1": 0, "LFJ2": 70, "LFJ3": 26, "LFJ4": 5, "LFJ5": 37}
# zero wrist
zero_wr: {"WRJ1": 0, "WRJ2": 0}
# north wrist
n_wr: {"WRJ1": 15, "WRJ2": 0}
# south wrist
s_wr: {"WRJ1": -30, "WRJ2": 0}
# east wrist
e_wr: {"WRJ1": 0, "WRJ2": 8}
# west wrist
w_wr: {"WRJ1": 0, "WRJ2": -14}
# northeast wrist
ne_wr: {"WRJ1": 15, "WRJ2": 8}
# northwest wrist
nw_wr: {"WRJ1": 15, "WRJ2": -14}
# southweast wrist
sw_wr: {"WRJ1": -30, "WRJ2": -14}
# southeast wrist
se_wr: {"WRJ1": -30, "WRJ2": 8}
# lateral lf ext side
l_ext_lf: {"LFJ4": -15}
# lateral rf ext side
l_ext_rf: {"RFJ4": -15}
# lateral mf ext side
l_ext_mf: {"MFJ4": 15}
# lateral ff ext side
l_ext_ff: {"FFJ4": 15}
# lateral all int side
l_int_all: {"FFJ4": -15, "MFJ4": -15, "RFJ4": 15, "LFJ4": 15}
# lateral all ext side
l_ext_all: {"FFJ4": 15, "MFJ4": 15, "RFJ4": -15, "LFJ4": -15}
# lateral ff int side
l_int_ff: {"FFJ4": -15}
# lateral mf int side
l_int_mf: {"MFJ4": -15}
# lateral rf int side
l_int_rf: {"RFJ4": 15}
# lateral lf int side
l_int_lf: {"LFJ4": 15}
# all zero
l_zero_all: {"FFJ4": 0, "MFJ4": 0, "RFJ4": 0, "LFJ4": 0}
# spock
l_spock: {"FFJ4": -20, "MFJ4": -20, "RFJ4": -20, "LFJ4": -20}
